Bug 269902 - [Curl] Add null check in WebSocketTaskCurl/NetworkDataTask/NetworkSession

kshukuwa opened the pull request:
#### 55dbfac13e1f1133bb160bae62cfaffed01659f3
<pre>
Bug 269902 - [Curl] Add null check in WebSocketTaskCurl/NetworkDataTask/NetworkSession
<a href="https://bugs.webkit.org/show_bug.cgi?id=269902">https://bugs.webkit.org/show_bug.cgi?id=269902</a>

Reviewed by Fujii Hironori.

m_channel.session() referenced in WebSocketTaskCurl is defined as WeakPtr&lt;NetworkSession&gt;.
Therefore, nullptr may be returned depending on the timing, so check for null before using them.

Also, NetworkSession::networkStorageSession() referenced by NetworkDataTaskCurl and NetworkSession
can return nullptr, so do a null check before using them.

* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
* Source/WebKit/NetworkProcess/curl/NetworkSessionCurl.cpp:
* Source/WebKit/NetworkProcess/curl/WebSocketTaskCurl.cpp:

Canonical link: <a href="https://commits.webkit.org/275304@main">https://commits.webkit.org/275304@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0faec52940e88d871de84d1b25fe5557a022a1b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41070 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20083 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43448 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43630 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37162 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23105 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17414 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33998 "Found 1 new test failure: fullscreen/exit-full-screen-video-crash.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41644 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17003 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35374 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14631 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14754 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44944 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37258 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36686 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40431 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13022 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38808 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17504 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17556 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5541 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17148 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
```diff
@@ -61,17 +61,18 @@ NetworkDataTaskCurl::NetworkDataTaskCurl(NetworkSession& session, NetworkDataTas
+        m_user = url.user();
+        m_password = url.password();
+        request.removeCredentials();
+        url = request.url();
```
Why do you do `url = request.url();` twice here?

kshukuwa replied:
This is to get the url after the credential has been removed with request.removeCredentials().

https://github.com/WebKit/WebKit/blob/74a570d11375304d6dd2fe7e2e2696e5f93bcff2/Source/WebCore/platform/network/ResourceRequestBase.cpp#L190-L199
=> `m_requestData.m_url.removeCredentials();` removes credentials in the URL.

https://github.com/WebKit/WebKit/blob/74a570d11375304d6dd2fe7e2e2696e5f93bcff2/Source/WebCore/platform/network/ResourceRequestBase.cpp#L121-L126
=> Then get the URL with the credential removed.

kshukuwa replied:
This process is exactly the same as the cocoa port. Also, the logic itself remains the same as before the change.

https://github.com/WebKit/WebKit/blob/74a570d11375304d6dd2fe7e2e2696e5f93bcff2/Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm#L196-L210

fujii replied:
Got it

fujii commented on Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
```diff
@@ -61,17 +61,18 @@ NetworkDataTaskCurl::NetworkDataTaskCurl(NetworkSession& session, NetworkDataTas
+        request.removeCredentials();
+        url = request.url();
+
+        if (auto* storageSession = m_session->networkStorageSession()) {
```
As per the new guildline, we shouldn't call methods without keeping a ref pointer.
https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines
> RefPtr storageSession = m_session->networkStorageSession()

kshukuwa replied:
NetworkSession::networkStorageSession() has nothing to do with RefPtr, but does it still need to be RefPtr?

https://github.com/WebKit/WebKit/blob/74a570d11375304d6dd2fe7e2e2696e5f93bcff2/Source/WebKit/NetworkProcess/NetworkSession.cpp#L90-L97

https://github.com/WebKit/WebKit/blob/74a570d11375304d6dd2fe7e2e2696e5f93bcff2/Source/WebKit/NetworkProcess/NetworkProcess.cpp#L563-L566

https://github.com/WebKit/WebKit/blob/74a570d11375304d6dd2fe7e2e2696e5f93bcff2/Source/WebKit/NetworkProcess/NetworkProcess.h#L549

fujii replied:
You are right. RefPtr isn't needed here.

kshukuwa committed:
Bug 269902 - [Curl] Add null check in WebSocketTaskCurl/NetworkDataTask/NetworkSession
https://bugs.webkit.org/show_bug.cgi?id=269902

Reviewed by Fujii Hironori.

m_channel.session() referenced in WebSocketTaskCurl is defined as WeakPtr<NetworkSession>.
Therefore, nullptr may be returned depending on the timing, so check for null before using them.

Also, NetworkSession::networkStorageSession() referenced by NetworkDataTaskCurl and NetworkSession
can return nullptr, so do a null check before using them.

* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
* Source/WebKit/NetworkProcess/curl/NetworkSessionCurl.cpp:
* Source/WebKit/NetworkProcess/curl/WebSocketTaskCurl.cpp:

Canonical link: https://commits.webkit.org/275304@main
